### DISCUSSION TITLE
Cancel signal not propagating to DataFetchers

### DISCUSSION THREAD

**Discussion by bclozel** [graphql-java/graphql-java#3691] [Q&A] [answered]
Assuming I have registered `DataFetcher` instances that return `CompletableFuture`. When starting the execution and cancelling in-flight:

```
CompletableFuture<ExecutionResult> result = graphQL.executeAsync(input);
result.cancel(true);
```

I'm not seeing the "cancel" signal being propagated up the chain and my `DataFetcher` is not cancelled as a result.
I think https://github.com/graphql-java/graphql-java/issues/75 is somewhat related, but I'm wondering if this is the expected behavior or if I can implement a custom component to achieve that goal. Is the message lost in the `AsyncExecutionStrategy`?

For background, Spring for GraphQL is trying to provide/document an interceptor for timeouts, but it seems that the "cancel" signal is lost along the way in `GraphQL`. This defeats a bit the purpose, as there is no way to cancel expensive calls if the execution has been cancelled.

**Comment by bbakerman** [comment, marked as answer]
I am not an expert in Future cancellation but I was under the impression that CompletableFuture does NOT propagate a signal upwards to stop a chained CF

re: elements from https://stackoverflow.com/questions/65539126/whats-the-intended-pattern-of-canceling-already-running-completablefutures

and 

https://nurkiewicz.com/2015/03/completablefuture-cant-be-interrupted.html

I did some local testing and chained CFs do not interrupt.

It would take a library like say https://github.com/vsilaev/tascalate-concurrent and its "dependent promises" for this to happen I suspect.

But if I am wrong here I am happy to be corrected and guided.

**Reply by bclozel** [reply]
Thanks for the links Brad, they're super useful.
Because the cancel signal flows from Reactor -> CompletableFuture and CompletableFuture -> Reactor, I wrongly assumed that this would work within a CompletableFuture chain.

I guess this limitation of CompletableFuture does not outweight the advantages it offers for graphql-java, so considering another library would be drastic at this point.

Thanks again!

**Reply by bbakerman** [reply]
I wish I knew about https://github.com/vsilaev/tascalate-concurrent 5 years ago.  I think graphql-java and dataloader would be based on dependant promises in retrospect

**Reply by bbakerman** [reply]
One crazy idea would be to keep a "list of DataFetcher" root CFs some where - so that they could be directly cancelled - this would interrupt them (modulo what they are doing - eg IO is ok but tight CPU work would not be)

In theory some one could write a Instrumentation to do this by "instrumentating" each DF and looking at its returned value (CF or not)

**Reply by bclozel** [reply]
Someone shared [a similar concept for Spring for GraphQL](https://github.com/spring-projects/spring-graphql/issues/450#issuecomment-2310152689). At this point, the overhead might be important for most cases and overall not really worth it?

**Reply by bbakerman** [reply]
Just spitballing :

There is a timeout mechanism on CompleteableFutures in Java 9 (which we are now on via min Java 11)

```
CompletableFuture<String> timedOut = asyncCF.orTimeout(5, TimeUnit.SECONDS);
```

This will throw a ` java.util.concurrent.TimeoutException` if the upstream CF does not complete

This is run via an internal thread pool (of 1) in the JDK

```
        static final ScheduledThreadPoolExecutor delayer;
        static {
            (delayer = new ScheduledThreadPoolExecutor(
                1, new DaemonThreadFactory())).
                setRemoveOnCancelPolicy(true);
        }
       ...
                   whenComplete(new Canceller(Delayer.delay(new Timeout(this),
                                                     timeout, unit)));
```

However it does not interrupt the original work.  Rather the "ticker" just returns an exception value for the CF - it does not stop any work that is happening - but one could imagine some how again instrumenting the DataFetcher CF so that it "cancels" on time out say just like we spoke about cancelling on some other signal (reactor back channel signal)

**Reply by bclozel** [reply]
We found a way to solve this in Spring for GraphQL for reactive data fetchers: https://github.com/spring-projects/spring-graphql/issues/1149

We are considering our options for non-reactive data fetchers here: https://github.com/spring-projects/spring-graphql/issues/1153
